### PR TITLE
Numpy Identity operator

### PR DESCRIPTION
**Pull request by hgt312**
## Description ##
Implementation of numpy `identity` operator in mxnet.
The version that cherry-pick on new numpy branch, old pr -> #15511.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- add op `identity`
- fix some typos/mistakes in doc of `zeros` and `ones`

## Comments ##
Welcome @reminisce, @haojin2, and others for reviewing.

### CODE REVIEW

**Comment by piyushghai** [comment]
@hgt312 Can you look into the CI failures ? 
@mxnet-label-bot Add [Numpy, operator]

**Comment by hgt312** [comment]
@piyushghai 
It a build error caused by mkl-dnn. There is a solution -> https://github.com/apache/incubator-mxnet/pull/15630/commits/128967b7e05ccc703d5a5165e802c0008103adbd.

File: python/mxnet/ndarray/numpy/_op.py [hunk 1]
```diff
@@ -170,14 +170,56 @@ def ones(shape, dtype=None, **kwargs):
...
+    _sanity_check_params('ones', ['order'], kwargs)
     ctx = kwargs.pop('ctx', current_context())
     if ctx is None:
         ctx = current_context()
     dtype = _np.float32 if dtype is None else dtype
     return _npi.ones(shape=shape, ctx=ctx, dtype=dtype, **kwargs)
 
 
+@set_module('mxnet.ndarray.numpy')
+def identity(n, dtype=None, **kwargs):
```
Comment (reminisce) [inline python/mxnet/ndarray/numpy/_op.py]: Change the interface to `identity(n, dtype=None, ctx=None)`.

File: python/mxnet/numpy/multiarray.py [hunk 2]
```diff
@@ -1497,6 +1497,40 @@ def ones(shape, dtype=None, **kwargs):
     return _mx_nd_np.ones(shape, dtype, **kwargs)
 
 
+@set_module('mxnet.numpy')
+def identity(n, dtype=None, **kwargs):
```
Comment (reminisce) [inline python/mxnet/numpy/multiarray.py]: Same as above.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 3]
```diff
@@ -972,14 +972,48 @@ def ones(shape, dtype=None, **kwargs):
...
+    _sanity_check_params('ones', ['order'], kwargs)
     ctx = kwargs.get('ctx', current_context())
     if ctx is None:
         ctx = current_context()
     dtype = _np.float32 if dtype is None else dtype
     return _npi.ones(shape=shape, ctx=ctx, dtype=dtype, **kwargs)
 
 
+@set_module('mxnet.symbol.numpy')
+def identity(n, dtype=None, **kwargs):
```
Comment (reminisce) [inline python/mxnet/symbol/numpy/_symbol.py]: Same as above.

File: src/operator/numpy/np_init_op.h [hunk 4]
```diff
@@ -107,6 +108,43 @@ void NumpyEyeFill(const nnvm::NodeAttrs& attrs,
   EyeFillImpl<xpu>(outputs[0], ctx, req, num_cols, param.N, param.k);
 }
 
+template<int req>
+struct identity {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* out_data, const int n) {
+    using namespace mxnet_op;
+
+    auto j = unravel(i, mshadow::Shape2(n, n));
```
Comment (reminisce) [inline src/operator/numpy/np_init_op.h]: Calculate j manually. `unravel` is slow.
```cpp
const index_t row_id = i / n;
const index_t col_id = i % n;
```

File: tests/python/unittest/test_numpy_ndarray.py [hunk 5]
```diff
@@ -148,6 +148,51 @@ def check_ones_array_creation(shape, dtype):
...
+@use_np
+def test_identity():
+    class TestIdentity(HybridBlock):
+        def __init__(self, shape, dtype=None):
+            super(TestIdentity, self).__init__()
+            self._n = n
+            self._dtype = dtype
+
+        def hybrid_forward(self, F, x, *args, **kwargs):
+            return x * F.np.identity(n, dtype)
```
Comment (reminisce) [inline tests/python/unittest/test_numpy_ndarray.py]: `dtype` -> `self._dtype`
`n` -> `self._n`

**Pull request closed without merging**